### ISSUE TITLE
[BUG] Memory overflow when lora fine-tuning meta-llama/Meta-Llama-3.1-8B-Instruct

### ISSUE THREAD

**Issue by hschaeufler** [ml-explore/mlx#1406]
**Describe the bug**
When finetuning a Llama 3.1 Instruct LLM, the memory consumption increases from 90 GB to over 200 GB after approx. 60 iterations. Because of this my M3 Max Macbook with 128GB memory crashes. Below is my config and the command. I cannot provide the data set itself as it contains sensitive data. But below is an example.

![image](https://github.com/user-attachments/assets/a2a3a5c3-c8e3-479f-8aca-538a02a4206b)

**To Reproduce**
Code Snippet:
```python
mlx_lm.lora --config "fine_tuning/lora_config.yaml"
```

```yaml
model: "meta-llama/Meta-Llama-3.1-8B-Instruct"
train: true
data: "fine_tuning/"
seed: 0
lora_layers: 16
batch_size: 4
iters: 2573
val_batches: 25
learning_rate: 1e-5
steps_per_report: 10
steps_per_eval: 200
resume_adapter_file: null
adapter_path: "adapters"
save_every: 100
test: false
test_batches: 100
max_seq_length: 8192
grad_checkpoint: false
use_dora: false
lora_parameters:
  keys: ["self_attn.q_proj", "self_attn.v_proj"]
  scale: 0.5
  rank: 16
  dropout: 0.05
lr_schedule:
  name: cosine_decay
  warmup: 60
  arguments: [1e-5, 62, 0] # passed to scheduler
```

Structure of train.jsonl and valid.jsonl (each line contains code from one class to test and one unit test)
```json
{"text":"### Code: import 'package:app\/features\/example\/domain\/entities\/entity.dart';\n... ### Test: import 'package:flutter_test\/flutter_test.dart';\n\nvoid main() {\n}\n"}
{"text":"### Code: import 'package:app\/features\/example\/domain\/entities\/entity.dart';\n... ### Test: import 'package:flutter_test\/flutter_test.dart';\n\nvoid main() {\n}\n"}
```

train.jsonl has 2058 records and valid.jsonl 257.

**Expected behavior**
Constant memory usage.

**Desktop (please complete the following information):**
 - OS Version: Version 14.6.1 (23G93)
 - MLX-LLM-Version: 0.18.1
 - Version: Mac15,9 with Apple M3 Max and 128GB Working Memory

**Additional context**
Found this: https://github.com/ml-explore/mlx-examples/issues/738 but memory checkpointing is deactivated.

**Comment by angeloskath** [comment]
It simply has to do with the size of the sequence length. Given that your maximum sequence length is 8192 I think you need to enable gradient checkpointing and possibly reduce the batch size.

With gradient checkpointing and a batch size of 1 I need 53GB of memory for the same 16 layer LoRA that you are doing while I need 134GB without checkpointing. One additional benefit is that with gradient checkpointing the memory required is disentangled from the number of lora layers so you can even fine-tune all layers with 55GB of RAM.

**Comment by angeloskath** [comment]
The following table is the memory required for sequence length of 8192.

```
Batch-size | Checkpointing | Memory
-----------+---------------+-------
  1        |  Yes          |  53 GB
  2        |  Yes          |  84 GB
  4        |  Yes          | 154 GB
  1        |   No          | 134 GB <-- with lora layers == 16
```

From the numbers above we can tell that your sequence length is actually not quite that big (at least all the time) and that is why the training starts and then the memory blows up.

I would
- enable gradient checkpointing
- reduce the batch size to 2
- possibly transform the model to float16 from bfloat16 to squeeze a tiny bit more speed

Feel free to close the issue or let me know if you think there is another problem.

**Comment by hschaeufler** [comment]
> The following table is the memory required for sequence length of 8192.
> 
> ```
> Batch-size | Checkpointing | Memory
> -----------+---------------+-------
>   1        |  Yes          |  53 GB
>   2        |  Yes          |  84 GB
>   4        |  Yes          | 154 GB
>   1        |   No          | 134 GB <-- with lora layers == 16
> ```
> 
> From the numbers above we can tell that your sequence length is actually not quite that big (at least all the time) and that is why the training starts and then the memory blows up.
> 
> I would
> 
> * enable gradient checkpointing
> * reduce the batch size to 2
> * possibly transform the model to float16 from bfloat16 to squeeze a tiny bit more speed
> 
> Feel free to close the issue or let me know if you think there is another problem.

Thank you very much, I will test that. Where can I view this table? Is there a command for this? The conversion is done via mlx_lm.convert [...] --dtype float16. 

A question about the Layers parameter. I have chosen 16 because it is the default value. In the Lora and Qlora paper and for other libraries I have not found any direct information or a recommendation. Do you have a literature recommendation or do you know where in the paper it is discussed?

Translated with www.DeepL.com/Translator (free version)

**Comment by hschaeufler** [comment]
I want to wait a little longer before closing the ticket because I'm interested in your opinion on my questions and wanted to give you the opportunity to respond. If you don't want to answer, you can also close the issue as resolve. Otherwise I'll set it to be solved at the beginning of next week.  Thank you so much for your help.

**Comment by angeloskath** [comment]
Sorry, for the late reply.

> Where can I view this table?

I just created it myself by making a dummy dataset with really long texts and set the max sequence length to 8192. I ran it on an M2 Ultra that's why it ran fine even with 154GB of RAM. It is actually quite interesting as all of these got >400 tokens per second even when using >150GB of RAM.

> The conversion is done via mlx_lm.convert [...] --dtype float16

Yes.

> A question about the Layers parameter.

There isn't a surefire approach here. The less layers you can get away with the better as it will be faster to train and require less memory. Using gradient checkpointing means you 'll need approximately as much memory as for a single layer but you still spend more time when finetuning all layers as the backward pass needs to run all the way back to the first trained layer.
In the paper they fine-tune all layers iirc. Unfortunately you will need to tune these parameters depending on your dataset and model.

Happy to answer more questions even if you post them in a closed issue ;-) .

**Comment by hschaeufler** [comment]
> There isn't a surefire approach here. The less layers you can get away with the better as it will be faster to train and require less memory. Using gradient checkpointing means you 'll need approximately as much memory as for a single layer but you still spend more time when finetuning all layers as the backward pass needs to run all the way back to the first trained layer.
> In the paper they fine-tune all layers iirc. Unfortunately you will need to tune these parameters depending on your dataset and model.

Thank you. :) I have found out that Meta-Llama-3.1-8B-Instruct has 32 layers and have now adjusted the parameter accordingly to 32. What do you mean that I have to adjust the parameter also to the data set? How does the data set influence the layer to finetune?

Maybe you can answer another question for me? Do you also have a recommendation regarding the Lora modules? In the Qlora paper a hyper parameter search was performed (https://arxiv.org/pdf/2305.14314) and they found out that LoRA r is unrelated to final performance if LoRA is used on all layers. Which is why I put all the modules I could find in the Llama Recipes. Do you have any experience/recommendations regarding model performance?

```
  keys: [
    "self_attn.q_proj",
    "self_attn.v_proj",
    "self_attn.k_proj",
    "self_attn.o_proj",
    "mlp.gate_proj",
    "mlp.down_proj",
    "mlp.up_proj"
  ]
```
